staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/convertlastapplied.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"encoding/json"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal"
+	openapiproto "k8s.io/kube-openapi/pkg/util/proto"
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+)
+
+type convertLastAppliedManager struct {
+	fieldManager Manager
+
+	typeConverter internal.TypeConverter
+	groupVersion  schema.GroupVersion
+}
+
+var _ Manager = &convertLastAppliedManager{}
+
+func NewConvertLastAppliedManager(fieldManager Manager, models openapiproto.Models, gv schema.GroupVersion) (Manager, error) {
+	typeConverter, err := internal.NewTypeConverter(models, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return &convertLastAppliedManager{
+		fieldManager:  fieldManager,
+		typeConverter: typeConverter,
+		groupVersion:  gv,
+	}, nil
+}
+
+// Update implements Manager.
+func (f *convertLastAppliedManager) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply takes field ownership of fields in the client-side apply annotation
+// so that upgrading from client-side apply to server-side apply doesn't conflict
+// by default because the user intends to manage the object with server-side apply now
+func (f *convertLastAppliedManager) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	// Upgrade the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager != "kubectl" {
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	}
+
+	lastAppliedFields, err := f.convertLastAppliedToFieldSet(liveObj)
+	// TODO: if there's an error, what do we do? should we remove the CSA annotation?
+	if err != nil {
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	}
+	fields := lastAppliedFields
+	existing, ok := managed.Fields()[manager]
+	if ok {
+		fields = existing.Set().Union(lastAppliedFields)
```

Comment:
I'm not completely sure yet that this is going to avoid conflicts ... you would have to remove the managed fields from  the previous manager too (we don't know who it is).

Reply:
Ah yes. Replaced with a second Apply, as discussed